File: Graph/__1000__/1200/1200.lua
local function Node(value)
    return {value = value, left = nil, right = nil}
end

local function insert(node, value)
    if node == nil then return Node(value) end
    if value < node.value then node.left = insert(node.left, value) end
    if value > node.value then node.right = insert(node.right, value) end
    return node
end

local function search(node, value)
    if node == nil then return false end
    if node.value == value then return true end
    if value < node.value then return search(node.left, value) end
    if value > node.value then return search(node.right, value) end
end

local first

local function PREFIX(node)
    if node == nil then return end

    if not first then io.write(" ") end
    io.write(string.char(node.value))
    first = false

    PREFIX(node.left)
    PREFIX(node.right)
end

local function INFIX(node)
    if node == nil then return end

    INFIX(node.left)

    if not first then io.write(" ") end
    io.write(string.char(node.value))
    first = false

    INFIX(node.right)
end

local function POSFIX(node)
    if node == nil then return end

    POSFIX(node.left)
    POSFIX(node.right)

    if not first then io.write(" ") end
    io.write(string.char(node.value))
    first = false
end


local root = nil -- raiz da árvore binaria.

while true do

    local input = io.read()
    if input == nil then break end

    if input == "PREFIXA" and root ~= nil then
        first = true
        PREFIX(root)
        io.write("\n")

    elseif input == "INFIXA" and root ~= nil then
        first = true
        INFIX(root)
        io.write("\n")

    elseif input == "POSFIXA" and root ~= nil then
        first = true
        POSFIX(root)
        io.write("\n")

    else

        local operation = input:sub(1, 1)
        local char = input:sub(3, 3)

        if operation == "I" then
            root = insert(root, string.byte(char))
        end

        if operation == "P" then
            local response = search(root, string.byte(char))
            io.write(char)

            if response then
                io.write(" existe\n")
            else
                io.write(" nao existe\n")
            end
        end
    end
end

File: Graph/__1000__/1200/1200.py
class Node:
    def __init__(self, char):
        self.char = char
        self.left = None
        self.right = None
        
class BST: #-> Binary Seaching Tree
    def __init__(self): self.root = None
    
    def infix(self): 
        vals = []
        self.infix_rec(self.root, vals)
        print(" ".join(vals))
    def infix_rec(self, node, vals):
        if node:
            self.infix_rec(node.left, vals)
            vals.append(node.char)
            self.infix_rec(node.right, vals)
            
    def prefix(self):
        vals = []
        self.prefix_rec(self.root, vals)
        print(" ".join(vals))
    def prefix_rec(self, node, vals):
        if node:
            vals.append(node.char)
            self.prefix_rec(node.left, vals)
            self.prefix_rec(node.right, vals)
            
    def posfix(self): 
        vals = []
        self.posfix_rec(self.root, vals)
        print(" ".join(vals))
    def posfix_rec(self, node, vals):
        if node:
            self.posfix_rec(node.left, vals)
            self.posfix_rec(node.right, vals)
            vals.append(node.char)
    
    def insert(self, char):self.root = self.add_rec(self.root, char)
    def add_rec(self, node, char):
        if node is None: return Node(char)
        elif ord(char) < ord(node.char): node.left = self.add_rec(node.left, char)
        elif ord(char) > ord(node.char): node.right = self.add_rec(node.right, char)
        return node
        
    def search(self, char): self.search_rec(self.root, char)
    def search_rec(self, node, char):
        if node is None: print(f"{char} nao existe")
        elif node.char == char: print(f"{char} existe")
        elif ord(char) < ord(node.char): self.search_rec(node.left, char)
        elif ord(char) > ord(node.char): self.search_rec(node.right, char)
        
class BST_Operations_I:
    @staticmethod
    def resolve():
        bst = BST()
        while True:
            try:
                request = input()
                parts = request.split()
                    
                if len(parts) == 2:
                    if parts[0] == "I": bst.insert(parts[1])
                    if parts[0] == "P": bst.search(parts[1])
                    
                if request == "INFIXA": bst.infix()
                if request == "PREFIXA": bst.prefix()
                if request == "POSFIXA": bst.posfix()
            except EOFError: break 
        
BST_Operations_I.resolve()